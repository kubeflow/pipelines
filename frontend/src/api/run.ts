// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export class RunMetadata {
  public id = '';
  public name = '';
  public namespace = '';
  public created_at = '';
  public scheduled_at = '';
  public status = '';

  public static buildFromObject(run: any): RunMetadata {
    const newRunMetadata = new RunMetadata();
    newRunMetadata.id = run.id;
    newRunMetadata.name = run.name;
    newRunMetadata.namespace = run.namespace;
    newRunMetadata.created_at = run.created_at;
    newRunMetadata.scheduled_at = run.scheduled_at;
    newRunMetadata.status = run.status;
    return newRunMetadata;
  }
}

export class Run {
  public run = RunMetadata.buildFromObject({});
  public workflow = '';

  public static buildFromObject(oldRun: any): Run {
    const newRun = new Run();
    newRun.run = RunMetadata.buildFromObject(oldRun.run);
    newRun.workflow = oldRun.workflow;
    return newRun;
  }
}
