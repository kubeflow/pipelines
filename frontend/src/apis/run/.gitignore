wwwroot/*.js
node_modules
typings

# Ignore git_push script
git_push.sh
