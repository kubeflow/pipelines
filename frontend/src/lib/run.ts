export interface RunStep {
  name: string;
  start: number;
  end: number;
  state: string;
  outputs: string;
}

export interface Run {
  end: Date;
  id: number;
  instanceId: number;
  parameterValues: { [key: string]: number | string };
  progress: number;
  start: Date;
  state: 'not started' | 'running' | 'errored' | 'completed';
  steps: RunStep[];
}
