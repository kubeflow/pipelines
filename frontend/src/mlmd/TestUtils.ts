/*
 * Copyright 2018,2021 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Artifact, Value } from 'src/third_party/mlmd';
import { ArtifactCustomProperties, ArtifactProperties } from './Api';

export const doubleValue = (number: number) => {
  const value = new Value();
  value.setDoubleValue(number);
  return value;
};

export const intValue = (number: number) => {
  const value = new Value();
  value.setIntValue(number);
  return value;
};

export const stringValue = (string: String) => {
  const value = new Value();
  value.setStringValue(String(string));
  return value;
};

export const buildTestModel = () => {
  const model = new Artifact();
  model.setId(1);
  model.setTypeId(1);
  model.setUri('gs://my-bucket/mnist');
  model.getPropertiesMap().set(ArtifactProperties.NAME, stringValue('test model'));
  model.getPropertiesMap().set(ArtifactProperties.DESCRIPTION, stringValue('A really great model'));
  model.getPropertiesMap().set(ArtifactProperties.VERSION, stringValue('v1'));
  model
    .getPropertiesMap()
    .set(ArtifactProperties.CREATE_TIME, stringValue('2019-06-12T01:21:48.259263Z'));
  model
    .getPropertiesMap()
    .set(
      ArtifactProperties.ALL_META,
      stringValue(
        '{"hyperparameters": {"early_stop": true, ' +
          '"layers": [10, 3, 1], "learning_rate": 0.5}, ' +
          '"model_type": "neural network", ' +
          '"training_framework": {"name": "tensorflow", "version": "v1.0"}}',
      ),
    );
  model
    .getCustomPropertiesMap()
    .set(ArtifactCustomProperties.WORKSPACE, stringValue('workspace-1'));
  model.getCustomPropertiesMap().set(ArtifactCustomProperties.RUN, stringValue('1'));
  return model;
};

export const testModel = buildTestModel();
