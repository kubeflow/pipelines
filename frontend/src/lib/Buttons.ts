/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import AddIcon from '@material-ui/icons/Add';
import CollapseIcon from '@material-ui/icons/UnfoldLess';
import ExpandIcon from '@material-ui/icons/UnfoldMore';
import { PageProps } from '../pages/Page';
import { URLParser } from './URLParser';
import { RoutePage, QUERY_PARAMS } from '../components/Router';
import { Apis } from './Apis';
import { errorToMessage, s } from './Utils';
import { ToolbarActionMap } from '../components/Toolbar';

export enum ButtonKeys {
  ARCHIVE = 'archive',
  CLONE_RUN = 'cloneRun',
  CLONE_RECURRING_RUN = 'cloneRecurringRun',
  RETRY = 'retry',
  COLLAPSE = 'collapse',
  COMPARE = 'compare',
  DELETE_RUN = 'deleteRun',
  DISABLE_RECURRING_RUN = 'disableRecurringRun',
  ENABLE_RECURRING_RUN = 'enableRecurringRun',
  EXPAND = 'expand',
  NEW_EXPERIMENT = 'newExperiment',
  NEW_RUN = 'newRun',
  NEW_RECURRING_RUN = 'newRecurringRun',
  NEW_RUN_FROM_PIPELINE = 'newRunFromPipeline',
  REFRESH = 'refresh',
  RESTORE = 'restore',
  TERMINATE_RUN = 'terminateRun',
  UPLOAD_PIPELINE = 'uploadPipeline',
}

export default class Buttons {
  private _map: ToolbarActionMap;
  private _props: PageProps;
  private _refresh: () => void;
  private _urlParser: URLParser;

  constructor(pageProps: PageProps, refresh: () => void, map?: ToolbarActionMap) {
    this._props = pageProps;
    this._refresh = refresh;
    this._urlParser = new URLParser(pageProps);
    this._map = map || {};
  }

  public getToolbarActionMap(): ToolbarActionMap {
    return this._map;
  }

  public archive(getSelectedIds: () => string[], useCurrentResource: boolean,
    callback: (selectedIds: string[], success: boolean) => void): Buttons {
    this._map[ButtonKeys.ARCHIVE] = {
      action: () => this._archive(getSelectedIds(), useCurrentResource, callback),
      disabled: !useCurrentResource,
      disabledTitle: useCurrentResource ? undefined : 'Select at least one resource to archive',
      id: 'archiveBtn',
      title: 'Archive',
      tooltip: 'Archive',
    };
    return this;
  }

  public cloneRun(getSelectedIds: () => string[], useCurrentResource: boolean): Buttons {
    this._map[ButtonKeys.CLONE_RUN] = {
      action: () => this._cloneRun(getSelectedIds()),
      disabled: !useCurrentResource,
      disabledTitle: useCurrentResource ? undefined : 'Select a run to clone',
      id: 'cloneBtn',
      style: { minWidth: 100 },
      title: 'Clone run',
      tooltip: 'Create a copy from this run\s initial state',
    };
    return this;
  }

  public cloneRecurringRun(getSelectedIds: () => string[], useCurrentResource: boolean): Buttons {
    this._map[ButtonKeys.CLONE_RECURRING_RUN] = {
      action: () => this._cloneRun(getSelectedIds(), true),
      disabled: !useCurrentResource,
      disabledTitle: useCurrentResource ? undefined : 'Select a recurring run to clone',
      id: 'cloneBtn',
      title: 'Clone recurring run',
      tooltip: 'Create a copy from this run\s initial state',
    };
    return this;
  }

  public retryRun(getSelectedIds: () => string[], useCurrentResource: boolean,
                  callback: (selectedIds: string[], success: boolean) => void): Buttons {
    this._map[ButtonKeys.RETRY] = {
      action: () => this._retryRun(getSelectedIds(), useCurrentResource, callback),
      disabled: !useCurrentResource,
      disabledTitle: useCurrentResource ? undefined : 'Select at least one resource to retry',
      id: 'retryBtn',
      title: 'Retry',
      tooltip: 'Retry',
    };
    return this;
  }

  public collapseSections(action: () => void): Buttons {
    this._map[ButtonKeys.COLLAPSE] = {
      action,
      icon: CollapseIcon,
      id: 'collapseBtn',
      title: 'Collapse all',
      tooltip: 'Collapse all sections',
    };
    return this;
  }

  public compareRuns(getSelectedIds: () => string[]): Buttons {
    this._map[ButtonKeys.COMPARE] = {
      action: () => this._compareRuns(getSelectedIds()),
      disabled: true,
      disabledTitle: 'Select multiple runs to compare',
      id: 'compareBtn',
      style: { minWidth: 125 },
      title: 'Compare runs',
      tooltip: 'Compare up to 10 selected runs',
    };
    return this;
  }

  public delete(getSelectedIds: () => string[], resourceName: 'pipeline' | 'recurring run config',
    callback: (selectedIds: string[], success: boolean) => void, useCurrentResource: boolean): Buttons {
    this._map[ButtonKeys.DELETE_RUN] = {
      action: () => resourceName === 'pipeline' ?
        this._deletePipeline(getSelectedIds(), useCurrentResource, callback) :
        this._deleteRecurringRun(getSelectedIds()[0], useCurrentResource, callback),
      disabled: !useCurrentResource,
      disabledTitle: useCurrentResource ? undefined : `Select at least one ${resourceName} to delete`,
      id: 'deleteBtn',
      title: 'Delete',
      tooltip: 'Delete',
    };
    return this;
  }

  public disableRecurringRun(getId: () => string): Buttons {
    this._map[ButtonKeys.DISABLE_RECURRING_RUN] = {
      action: () => this._setRecurringRunEnabledState(getId(), false),
      disabled: true,
      disabledTitle: 'Run schedule already disabled',
      id: 'disableBtn',
      title: 'Disable',
      tooltip: 'Disable the run\'s trigger',
    };
    return this;
  }

  public enableRecurringRun(getId: () => string): Buttons {
    this._map[ButtonKeys.ENABLE_RECURRING_RUN] = {
      action: () => this._setRecurringRunEnabledState(getId(), true),
      disabled: true,
      disabledTitle: 'Run schedule already enabled',
      id: 'enableBtn',
      title: 'Enable',
      tooltip: 'Enable the run\'s trigger',
    };
    return this;
  }

  public expandSections(action: () => void): Buttons {
    this._map[ButtonKeys.EXPAND] = {
      action,
      icon: ExpandIcon,
      id: 'expandBtn',
      title: 'Expand all',
      tooltip: 'Expand all sections',
    };
    return this;
  }

  public newExperiment(getPipelineId?: () => string): Buttons {
    this._map[ButtonKeys.NEW_EXPERIMENT] = {
      action: () => this._createNewExperiment(getPipelineId ? getPipelineId() : ''),
      icon: AddIcon,
      id: 'newExperimentBtn',
      outlined: true,
      style: { minWidth: 185 },
      title: 'Create experiment',
      tooltip: 'Create a new experiment',
    };
    return this;
  }

  public newRun(getExperimentId?: () => string): Buttons {
    this._map[ButtonKeys.NEW_RUN] = {
      action: () => this._createNewRun(false, getExperimentId ? getExperimentId() : undefined),
      icon: AddIcon,
      id: 'createNewRunBtn',
      outlined: true,
      primary: true,
      style: { minWidth: 130 },
      title: 'Create run',
      tooltip: 'Create a new run',
    };
    return this;
  }

  public newRunFromPipeline(getPipelineId: () => string): Buttons {
    this._map[ButtonKeys.NEW_RUN_FROM_PIPELINE] = {
      action: () => this._createNewRunFromPipeline(getPipelineId()),
      icon: AddIcon,
      id: 'createNewRunBtn',
      outlined: true,
      primary: true,
      style: { minWidth: 130 },
      title: 'Create run',
      tooltip: 'Create a new run',
    };
    return this;
  }

  public newRecurringRun(experimentId: string): Buttons {
    this._map[ButtonKeys.NEW_RECURRING_RUN] = {
      action: () => this._createNewRun(true, experimentId),
      icon: AddIcon,
      id: 'createNewRecurringRunBtn',
      outlined: true,
      style: { minWidth: 195 },
      title: 'Create recurring run',
      tooltip: 'Create a new recurring run',
    };
    return this;
  }

  public refresh(action: () => void): Buttons {
    this._map[ButtonKeys.REFRESH] = {
      action,
      id: 'refreshBtn',
      title: 'Refresh',
      tooltip: 'Refresh the list',
    };
    return this;
  }

  public restore(getSelectedIds: () => string[], useCurrentResource: boolean,
    callback: (selectedIds: string[], success: boolean) => void): Buttons {
    this._map[ButtonKeys.RESTORE] = {
      action: () => this._restore(getSelectedIds(), useCurrentResource, callback),
      disabled: !useCurrentResource,
      disabledTitle: useCurrentResource ? undefined : 'Select at least one resource to restore',
      id: 'restoreBtn',
      title: 'Restore',
      tooltip: 'Restore',
    };
    return this;
  }

  public terminateRun(getSelectedIds: () => string[], useCurrentResource: boolean,
    callback: (selectedIds: string[], success: boolean) => void): Buttons {
    this._map[ButtonKeys.TERMINATE_RUN] = {
      action: () => this._terminateRun(getSelectedIds(), useCurrentResource, callback),
      disabled: !useCurrentResource,
      disabledTitle: useCurrentResource ? undefined : 'Select at least one run to terminate',
      id: 'terminateRunBtn',
      title: 'Terminate',
      tooltip: 'Terminate execution of a run',
    };
    return this;
  }

  public upload(action: () => void): Buttons {
    this._map[ButtonKeys.UPLOAD_PIPELINE] = {
      action,
      icon: AddIcon,
      id: 'uploadBtn',
      outlined: true,
      style: { minWidth: 160 },
      title: 'Upload pipeline',
      tooltip: 'Upload pipeline',
    };
    return this;
  }

  private _cloneRun(selectedIds: string[], isRecurring?: boolean): void {
    if (selectedIds.length === 1) {
      const runId = selectedIds[0];
      let searchTerms;
      if (isRecurring) {
        searchTerms = {
          [QUERY_PARAMS.cloneFromRecurringRun]: runId || '',
          [QUERY_PARAMS.isRecurring]: '1'
        };
      } else {
        searchTerms = { [QUERY_PARAMS.cloneFromRun]: runId || '' };
      }
      const searchString = this._urlParser.build(searchTerms);
      this._props.history.push(RoutePage.NEW_RUN + searchString);
    }
  }

  private _retryRun(selectedIds: string[], useCurrent: boolean,
                    callback: (selectedIds: string[], success: boolean) => void): void {
    this._dialogActionHandler(
        selectedIds,
        'Retry this run?',
        useCurrent,
        id => Apis.runServiceApi.retryRun(id),
        callback,
        'Retry',
        'run'
    );
  }

  private _archive(selectedIds: string[], useCurrent: boolean,
    callback: (selectedIds: string[], success: boolean) => void): void {
    this._dialogActionHandler(
      selectedIds,
      `Run${s(selectedIds)} will be moved to the Archive section, where you can still view ` +
      `${selectedIds.length === 1 ? 'its' : 'their'} details. Please note that the run will not ` +
      `be stopped if it's running when it's archived. Use the Restore action to restore the ` +
      `run${s(selectedIds)} to ${selectedIds.length === 1 ? 'its' : 'their'} original location.`,
      useCurrent,
      id => Apis.runServiceApi.archiveRun(id),
      callback,
      'Archive',
      'run',
    );
  }

  private _restore(selectedIds: string[], useCurrent: boolean,
    callback: (selectedIds: string[], success: boolean) => void): void {
    this._dialogActionHandler(
      selectedIds,
      `Do you want to restore ${selectedIds.length === 1 ? 'this run to its' : 'these runs to their'} original location?`,
      useCurrent,
      id => Apis.runServiceApi.unarchiveRun(id),
      callback,
      'Restore',
      'run',
    );
  }

  private _deletePipeline(selectedIds: string[], useCurrentResource: boolean,
    callback: (selectedIds: string[], success: boolean) => void): void {
    this._dialogActionHandler(
      selectedIds,
      'Do you want to delete this Pipeline? This action cannot be undone.',
      useCurrentResource,
      id => Apis.pipelineServiceApi.deletePipeline(id),
      callback,
      'Delete',
      'pipeline',
    );
  }

  private _deleteRecurringRun(id: string, useCurrentResource: boolean,
    callback: (_: string[], success: boolean) => void): void {
    this._dialogActionHandler(
      [id],
      'Do you want to delete this recurring run config? This action cannot be undone.',
      useCurrentResource,
      jobId => Apis.jobServiceApi.deleteJob(jobId),
      callback,
      'Delete',
      'recurring run config',
    );
  }

  private _terminateRun(ids: string[], useCurrentResource: boolean,
    callback: (_: string[], success: boolean) => void): void {
    this._dialogActionHandler(
      ids,
      'Do you want to terminate this run? This action cannot be undone. This will terminate any'
      + ' running pods, but they will not be deleted.',
      useCurrentResource,
      id => Apis.runServiceApi.terminateRun(id),
      callback,
      'Terminate',
      'run',
    );
  }

  private _dialogActionHandler(selectedIds: string[], content: string, useCurrentResource: boolean,
    api: (id: string) => Promise<void>, callback: (selectedIds: string[], success: boolean) => void,
    actionName: string, resourceName: string): void {

    const dialogClosedHandler = (confirmed: boolean) =>
      this._dialogClosed(confirmed, selectedIds, actionName, resourceName, useCurrentResource, api, callback);

    this._props.updateDialog({
      buttons: [{
        onClick: async () => await dialogClosedHandler(false),
        text: 'Cancel',
      }, {
        onClick: async () => await dialogClosedHandler(true),
        text: actionName,
      }],
      content,
      onClose: async () => await dialogClosedHandler(false),
      title: `${actionName} ${useCurrentResource ? 'this' : selectedIds.length} ${resourceName}${useCurrentResource ? '' : s(selectedIds.length)}?`,
    });
  }

  private async _dialogClosed(confirmed: boolean, selectedIds: string[], actionName: string,
    resourceName: string, useCurrentResource: boolean, api: (id: string) => Promise<void>,
    callback: (selectedIds: string[], success: boolean) => void): Promise<void> {
    if (confirmed) {
      const unsuccessfulIds: string[] = [];
      const errorMessages: string[] = [];
      await Promise.all(selectedIds.map(async (id) => {
        try {
          await api(id);
        } catch (err) {
          unsuccessfulIds.push(id);
          const errorMessage = await errorToMessage(err);
          errorMessages.push(`Failed to ${actionName.toLowerCase()} ${resourceName}: ${id} with error: "${errorMessage}"`);
        }
      }));

      const successfulOps = selectedIds.length - unsuccessfulIds.length;
      if (useCurrentResource || successfulOps > 0) {
        this._props.updateSnackbar({
          message: `${actionName} succeeded for ${useCurrentResource ? 'this' : successfulOps} ${resourceName}${useCurrentResource ? '' : s(successfulOps)}`,
          open: true,
        });
        if (!useCurrentResource) {
          this._refresh();
        }
      }

      if (unsuccessfulIds.length > 0) {
        this._props.updateDialog({
          buttons: [{ text: 'Dismiss' }],
          content: errorMessages.join('\n\n'),
          title: `Failed to ${actionName.toLowerCase()} ${useCurrentResource ? '' : unsuccessfulIds.length + ' '}${resourceName}${useCurrentResource ? '' : s(unsuccessfulIds)}`,
        });
      }

      callback(unsuccessfulIds, !unsuccessfulIds.length);
    }
  }
  private _compareRuns(selectedIds: string[]): void {
    const indices = selectedIds;
    if (indices.length > 1 && indices.length <= 10) {
      const runIds = selectedIds.join(',');
      const searchString = this._urlParser.build({ [QUERY_PARAMS.runlist]: runIds });
      this._props.history.push(RoutePage.COMPARE + searchString);
    }
  }

  private _createNewExperiment(pipelineId: string): void {
    const searchString = pipelineId ? this._urlParser.build({
      [QUERY_PARAMS.pipelineId]: pipelineId
    }) : '';
    this._props.history.push(RoutePage.NEW_EXPERIMENT + searchString);
  }

  private _createNewRun(isRecurring: boolean, experimentId?: string): void {
    const searchString = this._urlParser.build(Object.assign(
      { [QUERY_PARAMS.experimentId]: experimentId || '' },
      isRecurring ? { [QUERY_PARAMS.isRecurring]: '1' } : {}));
    this._props.history.push(RoutePage.NEW_RUN + searchString);
  }

  private _createNewRunFromPipeline(pipelineId?: string): void {
    let searchString = '';
    const fromRunId = this._urlParser.get(QUERY_PARAMS.fromRunId);

    if (fromRunId) {
      searchString = this._urlParser.build(Object.assign(
        { [QUERY_PARAMS.fromRunId]: fromRunId }
      ));
    } else {
      searchString = this._urlParser.build(Object.assign(
        { [QUERY_PARAMS.pipelineId]: pipelineId || '' }
      ));
    }

    this._props.history.push(RoutePage.NEW_RUN + searchString);
  }

  private async _setRecurringRunEnabledState(id: string, enabled: boolean): Promise<void> {
    if (id) {
      const toolbarActions = this._props.toolbarProps.actions;

      // TODO(rileyjbauer): make sure this is working as expected
      const buttonKey = enabled ? ButtonKeys.ENABLE_RECURRING_RUN : ButtonKeys.DISABLE_RECURRING_RUN;

      toolbarActions[buttonKey].busy = true;
      this._props.updateToolbar({ actions: toolbarActions });
      try {
        await (enabled ? Apis.jobServiceApi.enableJob(id) : Apis.jobServiceApi.disableJob(id));
        this._refresh();
      } catch (err) {
        const errorMessage = await errorToMessage(err);
        this._props.updateDialog({
          buttons: [{ text: 'Dismiss' }],
          content: errorMessage,
          title: `Failed to ${enabled ? 'enable' : 'disable'} recurring run`,
        });
      } finally {
        toolbarActions[buttonKey].busy = false;
        this._props.updateToolbar({ actions: toolbarActions });
      }
    }
  }

}
