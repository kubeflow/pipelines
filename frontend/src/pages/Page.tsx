/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { ToolbarProps } from '../components/Toolbar';
import { BannerProps } from '../components/Banner';
import { SnackbarProps } from '@material-ui/core/Snackbar';
import { DialogProps } from '../components/Router';
import { errorToMessage } from '../lib/Utils';

export interface PageProps extends RouteComponentProps {
  toolbarProps: ToolbarProps;
  updateBanner: (bannerProps: BannerProps) => void;
  updateDialog: (dialogProps: DialogProps) => void;
  updateSnackbar: (snackbarProps: SnackbarProps) => void;
  updateToolbar: (toolbarProps: Partial<ToolbarProps>) => void;
}

export abstract class Page<P, S> extends React.Component<P & PageProps, S> {
  protected _isMounted = true;

  constructor(props: any) {
    super(props);
    this.props.updateToolbar(this.getInitialToolbarState());
  }

  public abstract render(): JSX.Element;

  public abstract getInitialToolbarState(): ToolbarProps;

  public abstract refresh(): Promise<void>;

  public componentWillUnmount(): void {
    this.clearBanner();
    this._isMounted = false;
  }

  public componentDidMount(): void {
    this.clearBanner();
  }

  public clearBanner(): void {
    this.props.updateBanner({});
  }

  public async showPageError(message: string, error?: Error, mode?: 'error' | 'warning'): Promise<void> {
    const errorMessage = await errorToMessage(error);
    this.props.updateBanner({
      additionalInfo: errorMessage ? errorMessage : undefined,
      message: message + (errorMessage ? ' Click Details for more information.' : ''),
      mode: mode || 'error',
      refresh: this.refresh.bind(this),
    });
  }

  public showErrorDialog(title: string, content: string): void {
    this.props.updateDialog({
      buttons: [{ text: 'Dismiss' }],
      content,
      title,
    });
  }

  protected setStateSafe(newState: Partial<S>, cb?: () => void): void {
    if (this._isMounted) {
      this.setState(newState as any, cb);
    }
  }
}
