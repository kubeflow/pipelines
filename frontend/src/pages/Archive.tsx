/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import RunList from './RunList';
import { Apis } from '../lib/Apis';
import { Page } from './Page';
import { RunStorageState } from '../apis/run';
import { ToolbarProps } from '../components/Toolbar';
import { classes } from 'typestyle';
import { commonCss, padding } from '../Css';
import { errorToMessage, s } from '../lib/Utils';

interface ArchiveState {
  selectedIds: string[];
}

export default class Archive extends Page<{}, ArchiveState> {
  private _runlistRef = React.createRef<RunList>();

  constructor(props: any) {
    super(props);

    this.state = {
      selectedIds: [],
    };
  }

  public getInitialToolbarState(): ToolbarProps {
    return {
      actions: [{
        action: () => this.props.updateDialog({
          buttons: [
            { onClick: async () => await this._unarchiveDialogClosed(true), text: 'Restore' },
            { onClick: async () => await this._unarchiveDialogClosed(false), text: 'Cancel' },
          ],
          onClose: async () => await this._unarchiveDialogClosed(false),
          title: `Restore ${this.state.selectedIds.length} resource${s(this.state.selectedIds.length)}?`,
        }),
        disabled: true,
        disabledTitle: 'Select at least one resource to restore',
        id: 'restoreBtn',
        title: 'Restore',
        tooltip: 'Restore',
      }, {
        action: this.refresh.bind(this),
        id: 'refreshBtn',
        title: 'Refresh',
        tooltip: 'Refresh the list of resources',
      }],
      breadcrumbs: [],
      pageTitle: 'Experiments',
    };
  }

  public render(): JSX.Element {
    return <div className={classes(commonCss.page, padding(20, 'lr'))}>
      <RunList onError={this.showPageError.bind(this)} selectedIds={this.state.selectedIds}
        onSelectionChange={this._selectionChanged.bind(this)} ref={this._runlistRef}
        storageState={RunStorageState.ARCHIVED} {...this.props} />
    </div>;
  }

  public async refresh(): Promise<void> {
    // Tell run list to refresh
    if (this._runlistRef.current) {
      this.clearBanner();
      await this._runlistRef.current.refresh();
    }
  }

  private _selectionChanged(selectedIds: string[]): void {
    const toolbarActions = [...this.props.toolbarProps.actions];
    // Unarchive button
    toolbarActions[0].disabled = !selectedIds.length;
    this.props.updateToolbar({ breadcrumbs: this.props.toolbarProps.breadcrumbs, actions: toolbarActions });
    this.setState({ selectedIds });
  }

  private async _unarchiveDialogClosed(confirmed: boolean): Promise<void> {
    if (confirmed) {
      const unsuccessfulIds: string[] = [];
      const errorMessages: string[] = [];
      await Promise.all(this.state.selectedIds.map(async (id) => {
        try {
          await Apis.runServiceApi.unarchiveRun(id);
        } catch (err) {
          unsuccessfulIds.push(id);
          const errorMessage = await errorToMessage(err);
          errorMessages.push(`Deleting run failed with error: "${errorMessage}"`);
        }
      }));

      const successfulObjects = this.state.selectedIds.length - unsuccessfulIds.length;
      if (successfulObjects > 0) {
        this.props.updateSnackbar({
          message: `Successfully restored ${successfulObjects} resource${s(successfulObjects)}!`,
          open: true,
        });
        this.refresh();
      }

      if (unsuccessfulIds.length > 0) {
        this.showErrorDialog(
          `Failed to restore ${unsuccessfulIds.length} resource${s(unsuccessfulIds)}`,
          errorMessages.join('\n\n'));
      }

      this._selectionChanged(unsuccessfulIds);
    }
  }

}
