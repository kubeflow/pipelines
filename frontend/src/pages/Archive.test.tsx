/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Archive from './Archive';
import TestUtils from '../TestUtils';
import { PageProps } from './Page';
import { RunStorageState } from '../apis/run';
import { ShallowWrapper, shallow } from 'enzyme';
import { ButtonKeys } from '../lib/Buttons';

describe('Archive', () => {
  const updateBannerSpy = jest.fn();
  const updateToolbarSpy = jest.fn();
  const historyPushSpy = jest.fn();
  let tree: ShallowWrapper;

  function generateProps(): PageProps {
    return TestUtils.generatePageProps(Archive, {} as any, {} as any, historyPushSpy,
      updateBannerSpy, null, updateToolbarSpy, null);
  }

  beforeEach(() => {
    updateBannerSpy.mockClear();
    updateToolbarSpy.mockClear();
    historyPushSpy.mockClear();
  });

  afterEach(() => tree.unmount());

  it('renders archived runs', () => {
    tree = shallow(<Archive {...generateProps()} />);
    expect(tree).toMatchSnapshot();
  });

  it('removes error banner on unmount', () => {
    tree = shallow(<Archive {...generateProps()} />);
    tree.unmount();
    expect(updateBannerSpy).toHaveBeenCalledWith({});
  });

  it('only enables restore button when at least one run is selected', () => {
    tree = shallow(<Archive {...generateProps()} />);
    TestUtils.flushPromises();
    tree.update();
    expect(TestUtils.getToolbarButton(updateToolbarSpy, ButtonKeys.RESTORE).disabled).toBeTruthy();
    tree.find('RunList').simulate('selectionChange', ['run1']);
    expect(TestUtils.getToolbarButton(updateToolbarSpy, ButtonKeys.RESTORE).disabled).toBeFalsy();
    tree.find('RunList').simulate('selectionChange', ['run1', 'run2']);
    expect(TestUtils.getToolbarButton(updateToolbarSpy, ButtonKeys.RESTORE).disabled).toBeFalsy();
    tree.find('RunList').simulate('selectionChange', []);
    expect(TestUtils.getToolbarButton(updateToolbarSpy, ButtonKeys.RESTORE).disabled).toBeTruthy();
  });

  it('refreshes the run list when refresh button is clicked', async () => {
    tree = shallow(<Archive {...generateProps()} />);
    const spy = jest.fn();
    (tree.instance() as any)._runlistRef = { current: { refresh: spy } };
    await TestUtils.getToolbarButton(updateToolbarSpy, ButtonKeys.REFRESH).action();
    expect(spy).toHaveBeenLastCalledWith();
  });

  it('shows a list of available runs', () => {
    tree = shallow(<Archive {...generateProps()} />);
    expect(tree.find('RunList').prop('storageState')).toBe(RunStorageState.ARCHIVED.toString());
  });
});
