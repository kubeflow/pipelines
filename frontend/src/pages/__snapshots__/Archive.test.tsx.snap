// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Archive renders archived runs 1`] = `
<div
  className="page"
>
  <RunList
    history={
      Object {
        "push": [MockFunction],
      }
    }
    location={Object {}}
    match={Object {}}
    onError={[Function]}
    onSelectionChange={[Function]}
    selectedIds={Array []}
    storageState="STORAGESTATE_ARCHIVED"
    toolbarProps={
      Object {
        "actions": Object {
          "refresh": Object {
            "action": [Function],
            "id": "refreshBtn",
            "title": "Refresh",
            "tooltip": "Refresh the list",
          },
          "restore": Object {
            "action": [Function],
            "disabled": true,
            "disabledTitle": "Select at least one resource to restore",
            "id": "restoreBtn",
            "title": "Restore",
            "tooltip": "Restore",
          },
        },
        "breadcrumbs": Array [],
        "pageTitle": "Archive",
      }
    }
    updateBanner={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
        ],
      }
    }
    updateDialog={null}
    updateSnackbar={null}
    updateToolbar={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "actions": Object {
                "refresh": Object {
                  "action": [Function],
                  "id": "refreshBtn",
                  "title": "Refresh",
                  "tooltip": "Refresh the list",
                },
                "restore": Object {
                  "action": [Function],
                  "disabled": true,
                  "disabledTitle": "Select at least one resource to restore",
                  "id": "restoreBtn",
                  "title": "Restore",
                  "tooltip": "Restore",
                },
              },
              "breadcrumbs": Array [],
              "pageTitle": "Archive",
            },
          ],
        ],
      }
    }
  />
</div>
`;
